Parser interface, implementation and Command class

kdkarki opened the pull request:
1. Parser interface created with single method parse.
2. GameParser class created as implementation of Parser interface
    a. The constructor accepts Dictionary object
    b. Stub methods created for different steps required in Parser to parse user input.
3. Command class created with 5 String fields.

gregwk commented:
Looks good! Here are some comments:

# Command class: For the time being, get rid of "errorCode" until we are sure we will need it. The client can assume that if errorMessage is the empty string, then no error exists.

# Parser interface: Looks good

# GameParser class:

(1) Make Dictionary and Grammar instance variables. They will be singletons, so you can do this in the following way:
Dictionary dictionary = GameDictionary.getInstance();
Grammar grammar = GameGrammar.getInstance();

(2) Make GameParser a singleton class:
private static GameParser instance = new GameParser();
private GameParser() { }
public static GameParser getInstance() { return instance; }

(3) I really like the way you are breaking this up into smaller methods. This is exactly what you need to do with the parser since the main "parse" method is quite complicated.

(4) In fact, these smaller methods are so relevant, I suggest we take some time to get their signatures correct, and then we make them public, so that we can test them individually. I'm am NOT suggesting we put them in the Parser interface -- that is fine and should only have one method -- but we will make them public in the GameParser class so that they can each be tested. Here are a few methods you will likely need (all except "match" is based on what you already have):

public String[] tokenize(String input);
public String[] removeStopWords(String[] stopWords, String[] words);
public void verifyWordsDefined(String[] words);
public void disambiguateNounWords(String[] nounWords);
public boolean match(String[] words, String pattern);

(5) To make sure information is retained between some of these calls, you will likely need to create some instance variables, along with public setters to they can be accessed in the JUnit test.

private Command command = new Command();
private final String[] stopWords = {"a", "an", "the"};
private String[] obj1NounWords;
private String[] obj2NounWords;

(6) All of the methods in (4) can potentially affect the "command" instance variable, because if there is an error, the "errorMessage" string of the command will be updated (and if !command.errorMessage.equals("") after one of these calls, then you can stop the parse routine and return the command).

(7) match will potentially update obj1NounWords and obj2NounWords.

In know this is a lot of information, but I think it will help testing if we do things this way. Note that the methods in (4) are likely incomplete, but I think they are a good start.

Dr. K

kdkarki commented:
Dr K, 

I have implemented the changes as per your suggestion. I have also started implementation of some of the methods start with tokenizeWords which breaks string input into array of strings. 

I have added instance variables of GameDictionary and GameGrammar though I think we should not do this and instead should inject any dependency (through setter). 

There are few TODO comments where I have questions and will bring them up in class/piazza.

Thanks,
Kshitij
